In [2]:
import pandas as pd
import numpy as np

<div><h1>1ª Tarefa</h1></div>

Primeiramente, tentei verificar se conseguiria trazer os arquivos .csv diretamente do site da PRF sem precisar baixá-los. Obviamente não consegui, mas não posso afirmar com certeza que é impossível fazer isso, pois não sei se esgotei todas as possibilidades. Portanto, pelo menos pra mim foi impossível. O fato é que pensei em uma forma que poderia ler os arquivos sem muita dor de cabeça, que foi baixando os arquivos compactados no site, extraindo eles e depois colocando eles em algum repositório no meu github e usando o link do arquivo gerado pelo github para poder ler eles no jupyter. Um detalhe é que só vim a pensar nisso apenas por conta de uma aula que tinha visto há um tempo sobre git que foi falado que arquivos como csv podem ser guardados tranquilamente em repositórios do GitHub. Ainda não havia usado muito o github como repositório, então aproveitei para dar uma brincada. 

In [3]:
df_18 = pd.read_csv('https://raw.githubusercontent.com/JVoliv/data-science-and-statistic/main/datatran2018.csv', sep = ';',encoding = 'ISO-8859-1')
df_19 = pd.read_csv('https://raw.githubusercontent.com/JVoliv/data-science-and-statistic/main/datatran2019.csv', sep = ';',encoding = 'ISO-8859-1') 
df_20 = pd.read_csv('https://raw.githubusercontent.com/JVoliv/data-science-and-statistic/main/datatran2020.csv', sep = ';',encoding = 'ISO-8859-1') 

Minha primeira tentativa de ler os arquivos .csv falhou por conta de um erro de codificação e para resolver tentei colocar encoding = 'utf-8', mas como quase sempre, deu errado e tentei novamente usando o ISO-8859-8. Deu certo mas a apresentação dos DF ficou feia, sem separação entre as colunas, e no caso do de 2019 ocorreu o erro "ParserError". Como já havia passado por isso antes sabia que era por conta do separador, e por isso coloquei o sep = ';'.

Abaixo estão as amostragens dos DF:

In [4]:
df_18.columns = df_18.columns.str.replace('[_]', ' ',regex=True)
df_18

id data inversa     dia semana   horario  uf     br     km  \
0       99973.0   2018-01-01  segunda-feira  00:20:00  RJ  116.0  303,5   
1       99976.0   2018-01-01  segunda-feira  00:40:00  SC  282.0    0,4   
2       99977.0   2018-01-01  segunda-feira  00:30:00  RJ  493.0      1   
3       99981.0   2018-01-01  segunda-feira  01:15:00  RS  386.0    134   
4       99982.0   2018-01-01  segunda-feira  00:20:00  RS  293.0  151,7   
...         ...          ...            ...       ...  ..    ...    ...   
69290  252256.0   2018-12-24  segunda-feira  09:15:00  PI  343.0  263,1   
69291  252652.0   2018-12-12   quarta-feira  07:00:00  BA  101.0    321   
69292  252801.0   2018-11-15   quinta-feira  10:48:00  SP  116.0    529   
69293  253040.0   2018-11-23    sexta-feira  19:30:00  RS  472.0  573,5   
69294  253078.0   2018-10-01  segunda-feira  04:31:00  GO   40.0      9   

                       municipio  \
0                        RESENDE   
1                  FLORIANOPOLIS   
2                       ITABORAI   
3                        SARANDI   
4                       CANDIOTA   
...                          ...   
69290                CAMPO MAIOR   
69291  PRESIDENTE TANCREDO NEVES   
69292             BARRA DO TURVO   
69293                 URUGUAIANA   
69294                   LUZIANIA   

                                          causa acidente  \
0                                      Condutor Dormindo   
1                     Não guardar distância de segurança   
2                                 Ultrapassagem Indevida   
3                                     Ingestão de Álcool   
4                            Falta de Atenção à Condução   
...                                                  ...   
69290                                   Agressão Externa   
69291                        Falta de Atenção à Condução   
69292                        Falta de Atenção à Condução   
69293  Desobediência às normas de trânsito pelo condutor   
69294                 Não guardar distância de segurança   

                      tipo acidente  ... feridos graves ilesos ignorados  \
0         Saída de leito carroçável  ...              4      1         0   
1                  Colisão traseira  ...              2      1         0   
2                   Colisão frontal  ...              0      3         1   
3               Colisão transversal  ...              0      2         0   
4         Saída de leito carroçável  ...              1      0         0   
...                             ...  ...            ...    ...       ...   
69290   Colisão com objeto estático  ...              0      1         0   
69291     Saída de leito carroçável  ...              0      0         0   
69292              Colisão traseira  ...              0      0         1   
69293  Queda de ocupante de veículo  ...              1      1         0   
69294     Saída de leito carroçável  ...              0      0         0   

      feridos veiculos      latitude     longitude  regional  delegacia  \
0           4        1     -22,46937     -44,44705     SR-RJ     DEL5/7   
1           2        2  -27,59971687  -48,57565694     SR-SC     DEL8/1   
2           1        3    -22,763901    -42,927532     SR-RJ     DEL5/4   
3           0        2    -27,953636    -52,916374     SR-RS    DEL9/14   
4           1        1    -31,395214  -53,78391174     SR-RS    DEL9/11   
...       ...      ...           ...           ...       ...        ...   
69290       0        1     -4,819326    -42,158821     SR-PI    DEL17/1   
69291       1        1  -13,44768518  -39,42066193     SR-BA    DEL10/5   
69292       1        2    -24,939211    -48,301248     SR-SP     DEL6/5   
69293       1        2  -29,76355779  -57,05200195     SR-RS    DEL9/13   
69294       1        1    -16,131158    -47,957739     SR-DF   UOP01/DF   

            uop  
0      UOP03/RJ  
1      UOP01/SC  
2      UOP02/RJ  
3      UOP01/RS  
4      UOP03/RS  
...         ...  
69290  UOP03/PI  


In [5]:
df_19.columns = df_19.columns.str.replace('[_]', ' ',regex=True)
df_19

id data inversa     dia semana   horario  uf     br     km  \
0      182210.0   2019-01-01    terça-feira  01:30:00  SP  116.0    218   
1      182211.0   2019-01-01    terça-feira  01:30:00  PR  373.0  177,3   
2      182212.0   2019-01-01    terça-feira  00:45:00  SC  101.0     16   
3      182214.0   2019-01-01    terça-feira  00:00:00  CE   20.0  429,5   
4      182215.0   2019-01-01    terça-feira  01:00:00  MG  381.0    327   
...         ...          ...            ...       ...  ..    ...    ...   
67441  266255.0   2019-10-07  segunda-feira  09:10:00  ES  101.0  270,5   
67442  266406.0   2019-11-25  segunda-feira  07:20:00  PR  116.0    125   
67443  266434.0   2019-06-03  segunda-feira  19:00:00  CE  222.0    1,4   
67444  266573.0   2019-07-13         sábado  19:35:00  PR  373.0    425   
67445  266627.0   2019-10-22    terça-feira  20:20:00  BA  324.0    608   

                municipio                       causa acidente  \
0               GUARULHOS          Falta de Atenção à Condução   
1            PONTA GROSSA          Falta de Atenção à Condução   
2                  GARUVA                     Animais na Pista   
3               FORTALEZA  Ingestão de Substâncias Psicoativas   
4                NOVA ERA          Falta de Atenção à Condução   
...                   ...                                  ...   
67441               SERRA          Falta de Atenção à Condução   
67442  FAZENDA RIO GRANDE          Falta de Atenção à Condução   
67443             CAUCAIA   Não guardar distância de segurança   
67444              CANDOI          Defeito Mecânico no Veículo   
67445        SIMOES FILHO                       Defeito na Via   

                      tipo acidente  ... feridos graves ilesos ignorados  \
0       Colisão com objeto estático  ...              0      0         0   
1                  Colisão traseira  ...              0      0         0   
2       Colisão com objeto estático  ...              0      0         0   
3       Colisão com objeto estático  ...              1      0         0   
4               Colisão transversal  ...              1      1         0   
...                             ...  ...            ...    ...       ...   
67441                    Tombamento  ...              1      1         0   
67442               Colisão lateral  ...              0      1         1   
67443              Colisão traseira  ...              1      1         0   
67444     Saída de leito carroçável  ...              1      0         0   
67445  Queda de ocupante de veículo  ...              0      0         0   

      feridos veiculos      latitude     longitude  regional  delegacia  \
0           4        1  -23,46052014  -46,48772478     SR-SP     DEL6/1   
1           1        1  -25,05533957  -50,22776753     SR-PR     DEL7/3   
2           1        1      -26,1216      -48,8826     SR-SC     DEL8/3   
3           1        1      -3,76999    -38,670063     SR-CE    DEL16/1   
4           1        2      -19,7609      -43,0306     SR-MG     DEL4/3   
...       ...      ...           ...           ...       ...        ...   
67441       1        2     -20,23107     -40,27387     SR-ES    DEL12/2   
67442       1        2     -25,59505     -49,31631     SR-PR     DEL7/1   
67443       1        2     -3,736507     -38,65337     SR-CE    DEL16/1   
67444       1        1      -25,7118     -52,18374     SR-PR     DEL7/3   
67445       1        1     -12,78156     -38,41656     SR-BA    DEL10/1   

            uop  
0      UOP01/SP  
1      UOP01/PR  
2      UOP01/SC  
3      UOP02/CE  
4      UOP01/MG  
...         ...  
67441  UOP01/ES  
67442  UOP03/PR  
67443  UOP01/CE  
67444  UOP02/PR  
67445  UOP01/BA  

[67446 rows x 30 columns]

In [6]:
df_20.columns = df_20.columns.str.replace('[_]', ' ',regex=True)
df_20

id data inversa    dia semana   horario  uf     br     km  \
0      260031.0   2020-01-01  quarta-feira  01:00:00  TO  153.0  678,1   
1      260036.0   2020-01-01  quarta-feira  01:00:00  RJ  116.0    178   
2      260037.0   2020-01-01  quarta-feira  01:52:00  SC  101.0  206,9   
3      260038.0   2020-01-01  quarta-feira  01:15:00  RO  364.0    236   
4      260039.0   2020-01-01  quarta-feira  01:30:00  PR  153.0    360   
...         ...          ...           ...       ...  ..    ...    ...   
63543  364615.0   2020-03-20   sexta-feira  09:00:00  RN  304.0  171,8   
63544  364810.0   2020-09-12        sábado  23:50:00  BA  116.0   11,5   
63545  364817.0   2020-10-30   sexta-feira  10:30:00  BA  116.0   14,3   
63546  364835.0   2020-09-20       domingo  23:00:00  BA  116.0      8   
63547  364903.0   2020-10-31        sábado  02:10:00  BA  116.0      3   

                  municipio                causa acidente  \
0                    GURUPI              Animais na Pista   
1               NOVA IGUACU  Falta de Atenção do Pedestre   
2                  SAO JOSE            Ingestão de Álcool   
3                    CACOAL       Velocidade Incompatível   
4                  REBOUCAS            Ingestão de Álcool   
...                     ...                           ...   
63543               ANGICOS   Falta de Atenção à Condução   
63544  VITORIA DA CONQUISTA            Ingestão de Álcool   
63545  VITORIA DA CONQUISTA   Falta de Atenção à Condução   
63546  VITORIA DA CONQUISTA       Velocidade Incompatível   
63547  VITORIA DA CONQUISTA              Animais na Pista   

                      tipo acidente  ... feridos graves ilesos ignorados  \
0           Atropelamento de Animal  ...              5      0         0   
1         Atropelamento de Pedestre  ...              0      1         0   
2                   Colisão lateral  ...              0      1         0   
3         Saída de leito carroçável  ...              1      0         0   
4                       Capotamento  ...              4      0         0   
...                             ...  ...            ...    ...       ...   
63543              Colisão traseira  ...              1      2         0   
63544               Colisão lateral  ...              0      2         0   
63545               Colisão lateral  ...              0      2         0   
63546  Queda de ocupante de veículo  ...              0      0         0   
63547       Atropelamento de Animal  ...              1      0         0   

      feridos veiculos      latitude     longitude  regional  delegacia  \
0           5        1  -11,77460203  -49,10744996     SR-TO   UOP01/TO   
1           0        1  -22,75223028   -43,4379103     SR-RJ     DEL5/1   
2           1        2  -27,59193546  -48,61824557     SR-SC     DEL8/1   
3           1        1  -11,44624577  -61,43761218     SR-RO    DEL21/2   
4           6        1  -25,67503796  -50,75089805     SR-PR     DEL7/3   
...       ...      ...           ...           ...       ...        ...   
63543       2        4   -5,68711053    -36,419458     SR-RN      UOP03   
63544       1        2  -14,87308214  -40,87141514     SR-BA      UOP01   
63545       2        2  -14,89597653  -40,86098671     SR-BA      UOP01   
63546       1        1  -14,84732305  -40,88634968     SR-BA      UOP01   
63547       1        1    -14,824848    -40,861311     SR-BA      UOP01   

            uop  
0           NaN  
1      UOP02/RJ  
2      UOP01/SC  
3      UOP02/RO  
4      UOP04/PR  
...         ...  
63543     DEL01  
63544     DEL08  
63545     DEL08  
63546     DEL08  
63547     DEL08  

[63548 rows x 30 columns]

Realizei a substituição dos underscore nos nomes das colunas por um espaço, apenas porque na minha opinão ficaria mais apresentável. 

<div><h1>2ª Tarefa</h1></div>

<div><h3>(a) quantos acidentes ocorreram em cada ano?</h3><div>

In [7]:
print(f"No ano de 2018 foram registrados {len(df_18)} acidentes.")

No ano de 2018 foram registrados 69295 acidentes.


In [8]:
print(f"No ano de 2019 foram registrados {len(df_19)} acidentes.")

No ano de 2019 foram registrados 67446 acidentes.


In [9]:
print(f"No ano de 2020 foram registrados {len(df_20)} acidentes.")

No ano de 2020 foram registrados 63548 acidentes.


Como as bases de dados dos anos requeridos tratam de acidentes, entendi que cada linha era um acidente. Por isso apenas fiz a contagem dos index de cada DF e apresentei dentro de um print, da forma como se pode ver acima. 

<div><h3>(b) quantas variáveis cada conjunto de dados registra?</h3><div>

In [10]:
print(f"A base de dados do ano de 2018 apresenta {len(df_18.columns)} variáveis")
print(f"A base de dados do ano de 2019 apresenta {len(df_19.columns)} variáveis")
print(f"A base de dados do ano de 2020 apresenta {len(df_20.columns)} variáveis")

A base de dados do ano de 2018 apresenta 30 variáveis
A base de dados do ano de 2019 apresenta 30 variáveis
A base de dados do ano de 2020 apresenta 30 variáveis


Nessa questão, como foi solicitado a QUANTIDADE das variáveis, decidi usar o mesmo método usado na letra A, só que dessa vez especificando que a contagem seria das colunas, usando o .columns. 

<div><h3>(c) quais as cinco cidades brasileiras onde mais ocorreram acidentes em rodovias federais?</h3><div>

In [11]:
df_18['municipio'].value_counts().head()

CURITIBA     1019
SAO JOSE      874
BRASILIA      867
GUARULHOS     755
PALHOCA       688
Name: municipio, dtype: int64

In [12]:
df_19['municipio'].value_counts().head()

CURITIBA     1094
BRASILIA     1090
SAO JOSE      819
GUARULHOS     717
SERRA         633
Name: municipio, dtype: int64

In [13]:
df_20['municipio'].value_counts().head()

BRASILIA     1040
CURITIBA      950
GUARULHOS     639
SAO JOSE      629
PALHOCA       593
Name: municipio, dtype: int64

Como nessa questão o que estava sendo pedido eram as 5 cidades que mais aparecem em cada DF, pensei primeiro em como poderia mostrar a quantidade que cada cidade aparece e para isso usei a função value_counts() que faz exatamente isso, mostra a quantidade de aparições de cada valor. E como só precisava dos dados referentes aos municípios, fiz a filtragem da coluna 'municipio', para que a contagem de ocorrência de cada item fosse feito somente nesta coluna. E por último coloquei o .head() pra mostrar as 5 primeiras, uma vez que a apresentação já é feita de forma a mostrar os maiores no começo da listagem. 

<div><h3>(d) quantos acidentes com feridos graves aconteceram na Paraíba em 2019?</h3><div>

In [14]:
acid_g = df_19['feridos graves']>0
acid_g

0        False
1        False
2        False
3         True
4         True
         ...  
67441     True
67442    False
67443     True
67444     True
67445    False
Name: feridos graves, Length: 67446, dtype: bool

Primeiramente eu criei uma variável para receber o resultado de uma filtragem que mostra True quando o valor na coluna 'feridos graves' é maior que zero. Acima coloquei para printar apenas para demonstrar isso.

In [15]:
cut = df_19[acid_g]
cut

id data inversa     dia semana   horario  uf     br     km  \
3      182214.0   2019-01-01    terça-feira  00:00:00  CE   20.0  429,5   
4      182215.0   2019-01-01    terça-feira  01:00:00  MG  381.0    327   
9      182223.0   2019-01-01    terça-feira  02:15:00  RS  386.0   73,7   
10     182225.0   2019-01-01    terça-feira  02:00:00  MG   50.0   77,3   
15     182231.0   2019-01-01    terça-feira  02:40:00  SC  282.0  599,3   
...         ...          ...            ...       ...  ..    ...    ...   
67436  266137.0   2019-12-23  segunda-feira  20:10:00  SE  235.0     57   
67439  266158.0   2019-01-15    terça-feira  05:00:00  AC  364.0    130   
67441  266255.0   2019-10-07  segunda-feira  09:10:00  ES  101.0  270,5   
67443  266434.0   2019-06-03  segunda-feira  19:00:00  CE  222.0    1,4   
67444  266573.0   2019-07-13         sábado  19:35:00  PR  373.0    425   

                   municipio  \
3                  FORTALEZA   
4                   NOVA ERA   
9      BOA VISTA DAS MISSOES   
10                UBERLANDIA   
15                CUNHA PORA   
...                      ...   
67436              ITABAIANA   
67439             RIO BRANCO   
67441                  SERRA   
67443                CAUCAIA   
67444                 CANDOI   

                                          causa acidente  \
3                    Ingestão de Substâncias Psicoativas   
4                            Falta de Atenção à Condução   
9                                     Ingestão de Álcool   
10     Desobediência às normas de trânsito pelo condutor   
15     Ingestão de álcool e/ou substâncias psicoativa...   
...                                                  ...   
67436                            Velocidade Incompatível   
67439                                     Defeito na Via   
67441                        Falta de Atenção à Condução   
67443                 Não guardar distância de segurança   
67444                        Defeito Mecânico no Veículo   

                      tipo acidente  ... feridos graves ilesos ignorados  \
3       Colisão com objeto estático  ...              1      0         0   
4               Colisão transversal  ...              1      1         0   
9                   Colisão frontal  ...              2      0         0   
10                  Colisão frontal  ...              2      1         0   
15      Colisão com objeto estático  ...              2      0         0   
...                             ...  ...            ...    ...       ...   
67436     Saída de leito carroçável  ...              1      0         1   
67439  Queda de ocupante de veículo  ...              1      0         0   
67441                    Tombamento  ...              1      1         0   
67443              Colisão traseira  ...              1      1         0   
67444     Saída de leito carroçável  ...              1      0         0   

      feridos veiculos    latitude   longitude  regional  delegacia       uop  
3           1        1    -3,76999  -38,670063     SR-CE    DEL16/1  UOP02/CE  
4           1        2    -19,7609    -43,0306     SR-MG     DEL4/3  UOP01/MG  
9           4        2  -27,693541  -53,300945     SR-RS      DEL08  UOP02/RS  
10          2        2    -18,9444    -48,2185     SR-MG    DEL4/15  UOP01/MG  
15          2        1  -26,820206  -53,155052     SR-SC       DEL7  UOP02/SC  
...       ...      ...         ...         ...       ...        ...       ...  
67436       1        1   -10,65001   -37,45125     SR-SE    DEL20/1  UOP02/SE  
67439       1        1   -10,01316   -67,71378     SR-AC   UOP01/AC       NaN  
67441       1        2   -20,23107   -40,27387     SR-ES    DEL12/2  UOP01/ES  
67443       1        2   -3,736507   -38,65337     SR-CE    DEL16/1  UOP01/CE  
67444       1        1    -25,7118   -52,18374     SR-PR     DEL7/3  UOP02/PR  

[14777 rows x 30 columns]

Então, criei outra variável para receber mais uma filtragem, dessa vez usei como filtro a variável que criei anteriormente. Isso para poder 'cortar' as linhas onde o número de feridos graves é zero (inclusive coloquei o nome da variável de 'cut', seguindo esse raciocínio.). Mais uma vez mostrei ela ao final apenas para fins demonstrativos. 

In [16]:
filt_pb = cut['uf']=='PB'
filt_pb

3        False
4        False
9        False
10       False
15       False
         ...  
67436    False
67439    False
67441    False
67443    False
67444    False
Name: uf, Length: 14777, dtype: bool

Mais uma vez fiz uma variável recendo mais uma filtragem booleana, para poder saber onde no DF "cut", na coluna "uf", aparece "PB".

In [17]:
result_final = cut[filt_pb]
print(f"A quantidade de acidentes graves no Estado da Paraíba é de {len(result_final)}.")

A quantidade de acidentes graves no Estado da Paraíba é de 414.


Por último fiz uma variável para receber a última filtragem que no caso é daquele mesmo DF "cut", recebendo como filtro justamente essa outra variável feita anteriormente para poder mostrar apenas as linhas onde o UF é PB. E depois disso apenas usei a função len() para poder obter quantidade. 

<div><h1>3ª Tarefa</h1></div>

<div><h3>(a) ranking de acidentes por estado para todos os anos (um só arquivo combinando a
informação dos três anos);</h3></div>

In [18]:
ac_18 = df_18['uf'].value_counts()
ac_19 = df_19['uf'].value_counts()
ac_20 = df_20['uf'].value_counts()

De início reutilizei um pouco do raciocínio usado no item C da 2ª tarefa, que foi: usar o .value_counts(). Só que dessa vez eu usei como parâmetro de filtragem de coluna o Estado e atribuí os 3 anos à variáveis.

In [19]:
data_ac = {'2018':ac_18, '2019':ac_19, '2020':ac_20}

Como nessa atribuição os resultados não saem diretamente como DF, para que eu pudesse apenas concatenar de alguma forma e organizar, eu aproveitei a situação pra me poupar um pouco de tempo e digitação. Eu coloquei as 3 series geradas em um dicionário, onde as chaves são os seu anos respectivos, e intitulei o dicionário de "data_ac".

In [20]:
rank_DF = pd.DataFrame(data_ac).rename_axis('UF').reset_index().sort_values(by = ['2018', '2019', '2020'], ascending=False, ignore_index=True)

Depois eu usei esse dicionário como dados de um novo DF. Infelizmente, mesmo tendo feito isso, ainda não está da forma como gostaria, pois fazendo somente "new_DF = pd.DataFrame(data)", vem com os Estados como índices, e eles deviam vir como colunas. Para consertar isso, primeiramente eu usei a função .rename_axis() para colocar um nome no index, para que ao trasformar esses index em colunas eles já estejam com o nome correto na coluna. Só após isso que usei o reset_index() para que os índices fossem redefinidos aos índices padrão. Tendo feito isso usei o ".sort_values(by = ['2018', '2019', '2020'], ascending=False, ignore_index=True)" para organizar o novo DF em formato de ranking, ou seja, com os maiores primeiro. Para isso, passei como parâmetro o "ascending=False", que faz com que seja mostrado na ordem que eu desejo, e o parâmetro "ignore_index=True", apenas para que os índices sejam mostrados de forma ordenada e fique um pouco mais organizado. 

In [21]:
rank_DF.to_csv('ranking-de-acidentes-2018-19-20.csv')

Por último, acima transformei em um arquivo .csv e abaixo segue a amostragem dos dados. 

In [22]:
rank_DF

UF  2018  2019  2020
0   MG  9066  8720  8374
1   SC  8497  8428  7220
2   PR  7952  7708  7190
3   RJ  4575  4606  4236
4   SP  4516  4377  4043
5   RS  4445  4595  4176
6   BA  3555  3471  3456
7   GO  3464  3354  3228
8   PE  2715  2736  2560
9   ES  2642  2842  2534
10  MT  2631  2137  2195
11  CE  1697  1694  1571
12  RO  1629  1533  1486
13  PB  1588  1564  1401
14  MS  1530  1435  1540
15  RN  1416  1437  1337
16  PI  1371  1346  1203
17  MA  1272  1075  1154
18  PA  1260   817   963
19  DF   867  1090  1040
20  AL   691   629   628
21  TO   589   520   595
22  SE   559   612   575
23  RR   236   237   250
24  AC   235   254   300
25  AP   173   128   155
26  AM   124   101   138

<div><h3>(b) acidentes por dia da semana para todos os anos (um só arquivo combinando a informação dos três anos).</h3></div>

In [23]:
wk_18 = df_18['dia semana'].value_counts()
wk_19 = df_19['dia semana'].value_counts()
wk_20 = df_20['dia semana'].value_counts()

data_dia = {'2018':wk_18, '2019':wk_19, '2020':wk_20}

wk_DF = pd.DataFrame(data_dia).rename_axis('UF').reset_index().sort_values(by = ['2018', '2019', '2020'], ascending=False, ignore_index=True)
wk_DF.to_csv('acidentes-por-dia-da-semana-2018-19-20.csv')
wk_DF

UF   2018   2019   2020
0        domingo  11283  11298  10420
1         sábado  11239  11099  10667
2    sexta-feira  10632  10381   9764
3  segunda-feira   9422   9134   8519
4   quinta-feira   9230   8755   8244
5   quarta-feira   8824   8374   8155
6    terça-feira   8665   8405   7779

Nessa questão eu usei o mesmo raciocínio da anterior, até reaproveitei o código, mudando apenas as variáveis.

<div><h1>4ª Tarefa</h1></div>

<div><h3>(a) qual a causa mais frequente e a mais rara de acidentes registrados nos três anos?</h3></div>

Nessa questão, como pede para averiguar as causas mais frequentes e as mais raras nos três anos, pesnsei em usar mais uma vez a mesma metodologia da questão anterior para que eu pudesse verificar isso. Inclusive mantive a forma de demonstração como se fosse um ranking para poder ver no topo de cada ano os acidentes mais recorrentes e no final os menos recorrentes. 

In [24]:
cs_18 = df_18['causa acidente'].value_counts()
cs_19 = df_19['causa acidente'].value_counts()
cs_20 = df_20['causa acidente'].value_counts()

data_causa = {'2018':cs_18, '2019':cs_19, '2020':cs_20}

causas_DF = pd.DataFrame(data_causa).rename_axis('Causas de Acidentes').reset_index().sort_values(by = ['2018', '2019', '2020'], ascending=False, ignore_index=True)
causas_DF

Causas de Acidentes     2018     2019   2020
0                         Falta de Atenção à Condução  25830.0  24992.0  22409
1   Desobediência às normas de trânsito pelo condutor   7189.0   8098.0   8802
2                             Velocidade Incompatível   6849.0   6028.0   5886
3                                  Ingestão de Álcool   5201.0   5419.0   5069
4                  Não guardar distância de segurança   4262.0   4214.0   3606
5                         Defeito Mecânico no Veículo   3716.0   3732.0   3742
6                                   Condutor Dormindo   2633.0   2483.0   2109
7                                  Pista Escorregadia   2549.0   2186.0   2017
8                        Falta de Atenção do Pedestre   2073.0   1971.0   1641
9                                    Animais na Pista   1614.0   1527.0   1410
10                             Ultrapassagem Indevida   1333.0   1173.0   1084
11                                     Defeito na Via   1126.0    973.0    960
12            Avarias e/ou desgaste excessivo no pneu   1011.0    799.0    902
13                                         Mal Súbito    677.0    715.0    584
14                          Restrição de Visibilidade    539.0    493.0    483
15           Objeto estático sobre o leito carroçável    539.0    437.0    446
16             Carga excessiva e/ou mal acondicionada    367.0    305.0    377
17                                   Agressão Externa    356.0    384.0    342
18                              Fenômenos da Natureza    310.0    230.0    221
19  Ingestão de álcool e/ou substâncias psicoativa...    297.0    394.0    357
20      Sinalização da via insuficiente ou inadequada    292.0    247.0    241
21  Desobediência às normas de trânsito pelo pedestre    273.0    392.0    388
22  Deficiência ou não Acionamento do Sistema de I...    186.0    160.0    187
23                Ingestão de Substâncias Psicoativas     73.0     94.0     86
24           Reação tardia ou ineficiente do condutor      NaN      NaN     24
25                     Ausência de reação do condutor      NaN      NaN     21
26  Condutor deixou de manter distância do veículo...      NaN      NaN     21
27                        Manobra de mudança de faixa      NaN      NaN     19
28                             Transitar na contramão      NaN      NaN     14
29  Acessar a via sem observar a presença dos outr...      NaN      NaN     13
30               Demais falhas mecânicas ou elétricas      NaN      NaN     13
31           Desrespeitar a preferência no cruzamento      NaN      NaN      8
32                   Ingestão de álcool pelo condutor      NaN      NaN      8
33                                              Chuva      NaN      NaN      6
34                                 Conversão proibida      NaN      NaN      6
35  Trafegar com motocicleta (ou similar) entre as...      NaN      NaN      5
36  Condutor desrespeitou a iluminação vermelha do...      NaN      NaN      4
37                           Pedestre andava na pista      NaN      NaN      4
38                                   Acesso irregular      NaN      NaN      3
39                               Falta de acostamento      NaN      NaN      3
40                           Transitar no acostamento      NaN      NaN      3
41                  Acumulo de água sobre o pavimento      NaN      NaN      2
42              Afundamento ou ondulação no pavimento      NaN      NaN      2
43                               Demais falhas na via      NaN      NaN      2
44                      Entrada inopinada do pedestre      NaN      NaN      2
45                              Iluminação deficiente      NaN      NaN      2
46                             Mal súbito do condutor      NaN      NaN      2
47                            Acostamento em desnível      NaN      NaN      1
48     Acumulo de areia ou detritos sobre o pavimento      NaN      NaN      1
49                  Acumulo de óleo sobre o pavimento      NaN      NaN      1
50   

- Causas de acidentes mais frequentes:

Nos 3 anos (2018, 2019 e 2020) o acidente mais recorrente foi o de "Falta de Atenção à Condução", com números de 25830 em 2018, 24992 em 2019 e 22409 em 2020.

- Causas de acidentes mais raras:

Nos anos de 2018 e 2019 o acidente mais raro foi o de "Ingestão de Substâncias Psicoativas", com número de 73 e 94 ocorrências registradas respectivamente.
Já no ano de 2020, muitos novos acidentes foram documentados. Acidentes esses que não tinham nos anos anteriores. E por isso aparecem tantos valores "NaN" nas colunas de 2018 e 2019. Assim, os mais raros do ano de 2020 são: Acostamento em desnível, Acumulo de areia ou detritos sobre o pavimento, Acumulo de óleo sobre o pavimento, Ausência de sinalização, Condutor usando celular, Curva acentuada, Estacionar ou parar em local proibido, Frear bruscamente, Fumaça, Ingestão de álcool ou de substâncias psicoativas, Pedestre cruzava a pista fora da faixa, Pista esburacada. 

Todas essas mais raras de 2020 são causas com apenas uma ocorrência documentada. 

<div><h3>(b) qual a proporção de pessoas ilesas e de feridos graves por mês em cada um dos anos
considerados? (proporção = numero de ilesos ou feridos graves/ número total de
    pessoas envolvidas no acidente)</h3><div>

Como nessa questão eu deveria apresentar as proporções por mês, pensei em usar a própria data que é apresentada no DF, mas para isso deveria primeiro transformar a coluna referente às datas para o formato de data, que pode ser visto abaixo a forma que usei isso. Tenho que admitir que tive que pesquisar um pouco pra saber como poderia fazer isso de forma eficaz e encontrei as respostas para as minhas dúvidas nos sites à seguir: https://ichi.pro/pt/trabalhando-com-data-e-hora-no-pandas-dataframe-112422831965575 e https://www.delftstack.com/pt/api/python-pandas/pandas-dataframe-dataframe.groupby-function/#:~:text=A%20função%20pandas.,o%20método%20grupo-por()%20. 

O uso do groupby foi feito para poder agrupar os dados por mês, então eu passei como parâmetro o DF referente à cada ano e suando o dt.month, para espescificar que seria por mês, claro, e retornei a soma de cada um. 

In [25]:
df_18['data inversa'] = pd.to_datetime(df_18['data inversa'])
ilesos_18 = df_18.groupby(df_18['data inversa'].dt.month)['ilesos'].sum()
graves_18 = df_18.groupby(df_18['data inversa'].dt.month)['feridos graves'].sum()
pessoas_18 = df_18.groupby(df_18['data inversa'].dt.month)['pessoas'].sum()

Tendo feito isso, apenas realizei a proporção da forma como é descrito no enunciado da questão, como está abaixo com o ano de 2018. Interessante apontar que são 12 linhas, pois cada uma é referente à um mês do ano. 

Após isso, simplemente fiz o mesmo para todos os outros anos. 

In [42]:
print("Proporção de ilesos de 2018, por mês:")
ilesos_18/pessoas_18

Proporção de ilesos de 2018, por mês:


data inversa
1     0.513291
2     0.498256
3     0.481535
4     0.430159
5     0.422230
6     0.425097
7     0.437509
8     0.427051
9     0.420665
10    0.430178
11    0.433356
12    0.432916
dtype: float64

In [27]:
print("Proporção de feridos graves de 2018, por mês:")
graves_18/pessoas_18

Proporção de feridos graves de 2018, por mês:


data inversa
1     0.090609
2     0.092191
3     0.101853
4     0.112774
5     0.114457
6     0.116395
7     0.110295
8     0.106371
9     0.115453
10    0.112645
11    0.115405
12    0.108103
dtype: float64

In [28]:
df_19['data inversa'] = pd.to_datetime(df_19['data inversa'])
ilesos_19 = df_19.groupby(df_19['data inversa'].dt.month)['ilesos'].sum()
graves_19 = df_19.groupby(df_19['data inversa'].dt.month)['feridos graves'].sum()
pessoas_19 = df_19.groupby(df_19['data inversa'].dt.month)['pessoas'].sum()

In [34]:
print("Proporção de ilesos de 2019, por mês:")
ilesos_19/pessoas_19

Proporção de ilesos de 2019, por mês:


data inversa
1     0.424224
2     0.424104
3     0.433193
4     0.415685
5     0.415162
6     0.423757
7     0.436815
8     0.413084
9     0.421064
10    0.418922
11    0.425812
12    0.422265
dtype: float64

In [35]:
print("Proporção de feridos graves de 2019, por mês:")
graves_19/pessoas_19

Proporção de feridos graves de 2019, por mês:


data inversa
1     0.113264
2     0.106789
3     0.114599
4     0.105970
5     0.120750
6     0.121617
7     0.116670
8     0.116683
9     0.110656
10    0.119805
11    0.111047
12    0.114405
dtype: float64

In [31]:
df_20['data inversa'] = pd.to_datetime(df_20['data inversa'])
ilesos_20 = df_20.groupby(df_20['data inversa'].dt.month)['ilesos'].sum()
graves_20 = df_20.groupby(df_20['data inversa'].dt.month)['feridos graves'].sum()
pessoas_20 = df_20.groupby(df_20['data inversa'].dt.month)['pessoas'].sum()

In [38]:
print("Proporção de ilesos de 2019, por mês:")
ilesos_20/pessoas_20

Proporção de ilesos de 2019, por mês:


data inversa
1     0.415565
2     0.420011
3     0.401021
4     0.376267
5     0.401330
6     0.423283
7     0.424321
8     0.423414
9     0.416016
10    0.411006
11    0.407042
12    0.412292
dtype: float64

In [37]:
print("Proporção de feridos graves de 2019, por mês:")
graves_20/pessoas_20

Proporção de feridos graves de 2019, por mês:


data inversa
1     0.107463
2     0.111356
3     0.112575
4     0.117949
5     0.115942
6     0.115936
7     0.108617
8     0.119168
9     0.117388
10    0.119716
11    0.120385
12    0.121993
dtype: float64

<div><h3>(c) mostre, reorganizando o conjunto de dados, se a pandemia conseguiu diminuir ou não
a incidência de acidentes nas rodovias federais.</h3></div>

Nessa questão decidi fazer de uma forma parecida com que tinha feito anteriormente, na tarefa 3. Pois, uma vez que precisava comparar mês a mês a quantidade de acidentes dos 3 anos para poder fazer as verificações solicitadas, precisaria de uma amostragem similar à que foi feita na outra questão. Por isso usei a mesma ideia mudando apenas a questão da ordenação, que no caso, precisaria ser mês a mês e por isso usei dt.month juntamente com a chamada de contagem de valores. 

In [53]:
new_18 = df_18['data inversa'].dt.month.value_counts()
new_19 = df_19['data inversa'].dt.month.value_counts()
new_20 = df_20['data inversa'].dt.month.value_counts()
df_final = {'2018':new_18,'2019':new_19, '2020':new_20}
total_ac = pd.DataFrame(df_final)
total_ac

2018  2019  2020
1   6940  5120  5497
2   6192  4963  5467
3   6684  5593  4790
4   5530  5436  3883
5   4818  5473  4657
6   5566  5578  4734
7   5431  5794  5127
8   5442  5703  5643
9   5448  5868  5689
10  5430  5881  6120
11  5460  5802  5745
12  6354  6235  6196

Como podemos ver acima, o ano de 2020, nos meses de Março até Junho, teve uma quantidade bem menor de acidentes em comparação com os anos de 2018 e 2019. Levando em consideração que esses meses foram os meses iniciais de pandemia e que o número de acidente nos demais meses seguem em uma quantidade similar às dos outros anos apresentados, pode-se dizer que por conta da pandemia os números de acidentes nas rodovias federais foram brevemente diminuidos. 